### PR TITLE
Add brod:fold/8 api

### PR DESCRIPTION
**Pull request by zmstone**

### CODE REVIEW

File: src/brod.erl [hunk 1]
```diff
@@ -226,6 +233,22 @@
 -type cg() :: #brod_cg{}.
 -type cg_protocol_type() :: binary().
 -type fetch_opts() :: kpro:fetch_opts().
+-type fold_acc() :: term().
+-type fold_fun() :: fun((message(), fold_acc()) ->
```
Comment (k32) [inline src/brod.erl]: Nit: Better use higher-kind type:

     -type fold_fun(Acc) :: fun((message(), Acc) -> {ok, Acc} | {error, any()}.

File: src/brod_utils.erl [hunk 2]
```diff
@@ -205,17 +207,59 @@ flatten_batches(BeginOffset, Header, Batches0) ->
...
+  case brod_client:get_leader_connection(Client, Topic, Partition) of
+    {ok, Conn} ->
+      %% do not close connection after use, managed by client
+      fold(Conn, Topic, Partition, Offset, Opts, Acc, Fun, Limits);
+    {error, Reason} ->
+      ?BROD_FOLD_RET(Acc, Offset, {error, Reason})
+  end;
+fold(Conn, Topic, Partition, Offset, Opts, Acc, Fun, Limits) ->
+  Fetch = make_fetch_fun(Conn, Topic, Partition, Opts),
+  Infinity = 1 bsl 64,
```
Comment (k32) [inline src/brod_utils.erl:258]: Good one!

File: src/brod_utils.erl [hunk 3]
```diff
@@ -478,6 +522,52 @@ get_stable_offset(Header) ->
...
+      do_acc(Spawn, Fetcher, NextOffset, NewAcc, Fun, Rest, End, Count - 1);
+    {error, Reason} ->
+      ok = kill_fetcher(Fetcher),
+      ?BROD_FOLD_RET(Acc, Offset, Reason)
+  end.
+
+kill_fetcher({Pid, Mref}) ->
+  exit(Pid, kill),
+  receive
+    {'DOWN', Mref, process, _, _} ->
```
Comment (k32) [inline src/brod_utils.erl]: Looks like this may hang forever if Pid crashed before. Add `after` clause?

File: changelog.md [hunk 4]
```diff
@@ -163,4 +163,9 @@
     belong to open transactions (transactions which are neigher committed nor aborted),
     may result in an empty message set, then `brod_consumer` or `brod_utils:fetch` jumps to
     the next offset (if it is less than high-watermark offset).
+* 3.8.0
+  * Bump to kafka_protocol 2.2.9 (allow `atom()` hostname)
+  * Add `brod:fold/8`. This API spawns a process to fetch-ahead while folding the prefiously
```
Comment (k32) [inline changelog.md]: Typo: s/prefiously/previously/

File: src/brod.erl [hunk 5]
```diff
@@ -766,14 +789,33 @@ fetch(ConnOrBootstrap, Topic, Partition, Offset) ->
...
 fetch(ConnOrBootstrap, Topic, Partition, Offset, Opts) ->
   brod_utils:fetch(ConnOrBootstrap, Topic, Partition, Offset, Opts).
 
+%% @doc Fold through messages in a partition.
+%% Works like `lists:foldl/2' but with below stop conditions:
+%% * Always returns after reach high watermark offset
+%% * Returns after the given message count limit is reached
+%% * OR returns after the given kafka offset is reached.
+%% * Returns if the `FoldFun' returns an `{error, Reason}' tuple.
+%% NOTE: Exceptions from evaluating `FoldFun' are not catught.
```
Comment (k32) [inline src/brod.erl]: Typo: s/catught/caught/

File: src/brod.erl [hunk 6]
```diff
@@ -766,14 +789,33 @@ fetch(ConnOrBootstrap, Topic, Partition, Offset) ->
...
+-spec fold(connection() | {[endpoint()], conn_config()} | client_id(),
```
Comment (k32) [inline src/brod.erl]: Nit: better use higher-kind type with `Acc` parameter.

File: src/brod_utils.erl (hunk 3 above)
Comment (zmstone) [reply src/brod_utils.erl]: You mean raise an exception in the `after` clause?
The pid is from `spawned_monitor`, there is no reason to miss a `DOWN` message unless eaten by the fold callback.

**Review by k32 — APPROVED** [review]

**Pull request merged**